### PR TITLE
[autoscaler] Fixing AWS instance types autofill

### LINKED ISSUE #11484: [autoscaler] EC2 instance type resources detection only works for first 100 instances
### What is the problem?

The AWS node provider's call to boto3.client("ec2").describe_instance_types() only returns 100 results. It seems we need to paginate through the full set of instance types, which exceeds 100: https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/ec2.html#EC2.Client.describe_instance_types

Otherwise, instances like "m4.16xlarge" are not found in the list: dict_keys(['m2.2xlarge', 'r5a.12xlarge', 'c6g.16xlarge', 'm6g.xlarge', 'r5dn.8xlarge', 'm5dn.12xlarge', 'c5n.2xlarge', 'm5ad.8xlarge', 'c4.8xlarge', 't2.2xlarge', 'm5d.large', 'd2.2xlarge', 'c6g.xlarge', 'm5d.8xlarge', 'r5ad.24xlarge', 'inf1.xlarge', 'c5d.12xlarge', 't4g.nano', 'z1d.2xlarge', 't1.micro', 'r5a.xlarge', 'c5ad.16xlarge', 'c5ad.12xlarge', 'c5d.large', 'r6g.8xlarge', 'm5a.2xlarge', 'c5.4xlarge', 'c3.large', 'm5a.4xlarge', 'f1.2xlarge', 'r3.large', 'm5.24xlarge', 'x1e.8xlarge', 'c5ad.24xlarge', 'm5.8xlarge', 'c6gd.metal', 'r5ad.8xlarge', 'r5ad.12xlarge', 'm5a.16xlarge', 'm5n.16xlarge', 'm6g.metal', 'r6g.medium', 'm5d.16xlarge', 't3a.small', 'm6gd.metal', 'r4.16xlarge', 'a1.medium', 'i3en.3xlarge', 'c4.4xlarge', 'm2.4xlarge', 'r6gd.2xlarge', 'i3en.12xlarge', 'r3.8xlarge', 'r6gd.large', 'x1e.32xlarge', 'm5ad.4xlarge', 'm5dn.large', 'r6gd.xlarge', 'g3.16xlarge', 'm5dn.2xlarge', 'm5dn.8xlarge', 't3.2xlarge', 'g4dn.metal', 'r5.24xlarge', 'r6gd.medium', 't3a.2xlarge', 'm6gd.xlarge', 'c5ad.4xlarge', 'c5a.4xlarge', 't4g.medium', 'm5ad.2xlarge', 'i2.2xlarge', 'm4.4xlarge', 'm5ad.large', 'm5dn.xlarge', 'r6g.16xlarge', 'c3.8xlarge', 'm6gd.12xlarge', 'm5.12xlarge', 'r5a.large', 'm5ad.12xlarge', 'm5.xlarge', 'i3en.24xlarge', 'm5d.xlarge', 't3.xlarge', 'x1.16xlarge', 'm6gd.4xlarge', 'r5a.2xlarge', 'c5n.large', 'm6gd.medium', 'c5n.4xlarge', 'x1e.2xlarge', 'r5n.4xlarge', 'c6g.12xlarge', 'r5n.12xlarge', 'm5d.metal', 'd2.8xlarge', 'm5.metal', 'r6g.large', 'c5.xlarge'])

Btw, we should also be raising an error if an instance type is not found in this dict, instead of silently ignoring it.

### Reproduction (REQUIRED)

Try to create a cluster with m4.16xl instance type--- resources will not be auto filled.

### PR DESCRIPTION
**Pull request by AmeerHajAli**
This fixes the issues that instance type autofill only filled for the first 100 instances returned by boto3. 
I manually tested this and compared every single instance returned versus the available instances on my AWS account.
for example, in `us-west-2`, there are 340 instance types and this actually returns 340 instances. Here is an example output for the current `aws/example-multi-node-type.yaml`:
```
Updating the resources of cpu_4_ondemand to {'CPU': 4}.
Updating the resources of cpu_16_spot to {'CPU': 16, 'Custom1': 1, 'is_spot': 1}.
Updating the resources of gpu_1_ondemand to {'CPU': 4, 'GPU': 1, 'accelerator_type:K80': 1, 'Custom2': 2}.
Updating the resources of gpu_8_ondemand to {'CPU': 32, 'GPU': 4, 'accelerator_type:V100': 1}.
```

Also now that we are comfortable it covers all the available node types in a region, I made it throw an error if the instance type was not found.

## Related issue number
Closes #11484 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by wuisawesome — APPROVED** [review]
lgtm, out of curiosity, would it be possible/easy to test this?

**Review by ericl — APPROVED** [review]

File: python/ray/autoscaler/_private/util.py [hunk 1]
```diff
@@ -107,8 +107,8 @@ def fillout_defaults(config: Dict[str, Any]) -> Dict[str, Any]:
     except Exception:
         # We don't want to introduce new errors with filling available node
         # types resources feature.
-        logger.exception("Failed to autodetect node resources")
-
+        logger.exception("Failed to autodetect node resources.")
+        raise
```
Comment (wuisawesome) [inline python/ray/autoscaler/_private/util.py:111]: Actually, didn't we explicitly not re-raise this exception before? Why re-raise it now?

Comment (AmeerHajAli) [reply python/ray/autoscaler/_private/util.py:111]: Didn't we agree that the instance should be available in the instance types? And if a value error was raised the used should fix the node type.

Comment (AmeerHajAli) [reply python/ray/autoscaler/_private/util.py:109]: > lgtm, out of curiosity, would it be possible/easy to test this?

BTW, I realized this already tests it:
`python/ray/tests/test_autoscaler.py::AutoscalingTest::testValidateDefaultConfig`

**Pull request merged**